src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -468,50 +451,68 @@ private unsafe void InnerStartOperationReceiveMessageFrom()
                 _wsaRecvMsgWSABufferArray[0].Pointer = _ptrSingleBuffer;
                 _wsaRecvMsgWSABufferArray[0].Length = _count;
                 wsaRecvMsgWSABufferArray = _wsaRecvMsgWSABufferArray;
+                wsaRecvMsgWSABufferCount = 1;
             }
             else
             {
                 // Use the multi-buffer WSABuffer.
                 wsaRecvMsgWSABufferArray = _wsaBufferArray;
+                wsaRecvMsgWSABufferCount = (uint)_bufferListInternal.Count;
             }
 
             // Ensure the array is pinned.
+            Debug.Assert(!_wsaRecvMsgWSABufferArrayGCHandle.IsAllocated || _wsaRecvMsgWSABufferArrayGCHandle.Target == wsaRecvMsgWSABufferArray);
             if (!_wsaRecvMsgWSABufferArrayGCHandle.IsAllocated)
             {
                 _wsaRecvMsgWSABufferArrayGCHandle = GCHandle.Alloc(wsaRecvMsgWSABufferArray, GCHandleType.Pinned);
-                _ptrWSARecvMsgWSABufferArray = Marshal.UnsafeAddrOfPinnedArrayElement(wsaRecvMsgWSABufferArray, 0);
-            }
-            else
-            {
-                Debug.Assert(_wsaRecvMsgWSABufferArrayGCHandle.Target == wsaRecvMsgWSABufferArray);
-                Debug.Assert(_ptrWSARecvMsgWSABufferArray == Marshal.UnsafeAddrOfPinnedArrayElement(wsaRecvMsgWSABufferArray, 0));
             }
 
             // Fill in WSAMessageBuffer.
             unsafe
             {
-                Interop.Winsock.WSAMsg* pMessage = (Interop.Winsock.WSAMsg*)_ptrWSAMessageBuffer; ;
-                pMessage->socketAddress = _ptrSocketAddressBuffer;
+                Interop.Winsock.WSAMsg* pMessage = (Interop.Winsock.WSAMsg*)PtrWSAMessageBuffer;
+                pMessage->socketAddress = PtrSocketAddressBuffer;
                 pMessage->addressLength = (uint)_socketAddress.Size;
-                pMessage->buffers = _ptrWSARecvMsgWSABufferArray;
-                if (_buffer != null)
+                fixed (void* ptrWSARecvMsgWSABufferArray = &wsaRecvMsgWSABufferArray[0])
```

Comment:
> `UnsafeAddrOfPinnedArrayElement`, `GCHandle.AddrOfPinnedObject`

All these methods have argument validation, etc. that makes them slower than necessary.

`pMessage->buffers = (IntPtr)Unsafe.AsPointer(ref wsaRecvMsgWSABufferArray[0])` should be the fastest way to do this (a few instructions better than the dummy fixed statement), but it would require adding a dependency on `S.R.CS.Unsafe`.